<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# Проект «Анализ вакансий из HeadHunter»
   

In [1]:
import pandas as pd
import psycopg2

In [3]:
import warnings
warnings.filterwarnings('ignore')

In [ ]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

## 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в базе (вакансии находятся в таблице `vacancies`).

In [14]:
# текст запроса
query_3_1 = f'''
SELECT COUNT(*) AS vacancies_count
FROM vacancies;
'''

In [15]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

vacancies_count
0            49197

2. Напишите запрос, который посчитает количество работодателей (таблица `employers`).

In [ ]:
# текст запроса
query_3_2 = '''
SELECT COUNT(*) AS employers_count
FROM employers;
'''

In [17]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
df

employers_count
0            23501

3. Посчитайте с помощью запроса количество регионов (таблица `areas`).

In [ ]:
# текст запроса
query_3_3 = '''
SELECT COUNT(*) AS areas_count
FROM areas;
'''

In [19]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
df

areas_count
0         1362

4. Посчитайте с помощью запроса количество сфер деятельности в базе (таблица `industries`).

In [ ]:
# текст запроса
query_3_4 = '''
SELECT COUNT(*) AS industries_count
FROM industries;
'''

In [22]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
df

industries_count
0               294

***

### Выводы по предварительному анализу данных
Данные в базе достаточно объемные и разнообразные, что позволяет проводить статистически значимый анализ: представлено 49197 вакансий, 23501 работодатель, 1362 региона и 294 отрасли.

## 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (`cnt`) вакансий в каждом регионе (`area`).
Отсортируйте по количеству вакансий в порядке убывания.

In [ ]:
# текст запроса
query_4_1 = '''
SELECT 
    a.name AS area,
    COUNT(v.id) AS cnt
FROM vacancies v
JOIN areas a ON v.area_id = a.id
GROUP BY a.name
ORDER BY cnt DESC;
'''

In [24]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df

area   cnt
0             Москва  5333
1    Санкт-Петербург  2851
2              Минск  2112
3        Новосибирск  2006
4             Алматы  1892
..               ...   ...
764       Тарко-Сале     1
765    Новоаннинский     1
766            Бирск     1
767           Сасово     1
768             Анна     1

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [ ]:
# текст запроса
query_4_2 = '''
SELECT 
    COUNT(*) AS vacancies_with_salary
FROM vacancies
WHERE salary_from IS NOT NULL OR salary_to IS NOT NULL;
'''

In [26]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
df

vacancies_with_salary
0                  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до **целого числа**.

In [ ]:
# текст запроса
query_4_3 = '''
SELECT 
    ROUND(AVG(salary_from)) AS avg_salary_from,
    ROUND(AVG(salary_to)) AS avg_salary_to
FROM vacancies
'''

In [28]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
df

avg_salary_from  avg_salary_to
0          71065.0       110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (`schedule`) и типа трудоустройства (`employment`), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [ ]:
# текст запроса
query_4_4 = '''
SELECT 
    schedule,
    employment,
    COUNT(*) AS cnt
FROM vacancies
GROUP BY schedule, employment
ORDER BY cnt DESC;
'''

In [30]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
df

schedule           employment    cnt
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля «Требуемый опыт работы» (`experience`) в порядке возрастания количества вакансий, в которых указан данный вариант опыта.

In [ ]:
# текст запроса
query_4_5 = '''
SELECT 
    experience,
    COUNT(*) AS cnt
FROM vacancies
GROUP BY experience
ORDER BY cnt ASC;
'''

In [32]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df

experience    cnt
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

### Выводы по детальному анализу вакансий
- Большинство вакансий сосредоточено в крупных городах, таких как Москва и Санкт-Петербург, при этом встречаются регионы с единичными вакансиями, что говорит о высокой концентрации рынка труда в мегаполисах.

- Примерно у половины вакансий указана хотя бы одна граница зарплаты, средняя нижняя граница составляет 71065, а верхняя - 110537.

- По графику работы и типу занятости видно, что полный рабочий день с полной занятостью это наиболее распространенный формат, за ним следуют удаленная работа и гибкий график. Нестандартные форматы, такие как проектная работа или стажировка, встречаются реже.

- Большинство вакансий рассчитаны на специалистов с 1-6 лет опыта, но также достаточно предложений для соискателей без опыта, что говорит о разнообразии требований к квалификации.

## 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [ ]:
# текст запроса
query_5_1 = '''
WITH ranked_employers AS (
    SELECT 
        e.name AS employer,
        COUNT(v.id) AS vacancies_count,
        ROW_NUMBER() OVER (ORDER BY COUNT(v.id) DESC) AS rn
    FROM vacancies v
    JOIN employers e ON v.employer_id = e.id
    GROUP BY e.name
)
SELECT employer, vacancies_count
FROM ranked_employers
WHERE rn IN (1, 5);
'''

In [34]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df

employer  vacancies_count
0         Яндекс             1933
1  Газпром нефть              331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [ ]:
# текст запроса
query_5_2 = '''
WITH employers_per_area AS (
       SELECT 
              e.area AS area_id,
              COUNT(*) AS employers_count
         FROM employers AS e
        GROUP BY e.area
),
vacancies_per_area AS (
       SELECT 
              v.area_id,
              COUNT(*) AS vacancies_count
         FROM vacancies AS v
        GROUP BY v.area_id
),
area_stats AS (
       SELECT 
              a.id AS area_id,
              a.name AS area_name,
              COALESCE(emp.employers_count, 0) AS employers_count,
              COALESCE(vac.vacancies_count, 0) AS vacancies_count
         FROM areas AS a
         LEFT JOIN employers_per_area AS emp
           ON emp.area_id = a.id
         LEFT JOIN vacancies_per_area AS vac
           ON vac.area_id = a.id
)
SELECT 
       area_name,
       employers_count,
       vacancies_count
  FROM area_stats
 ORDER BY vacancies_count ASC, employers_count DESC;
'''

In [6]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
df

area_name  employers_count  vacancies_count
0                 Россия              410                0
1              Казахстан              207                0
2     Московская область               75                0
3     Краснодарский край               19                0
4     Ростовская область               18                0
...                  ...              ...              ...
1357              Алматы              721             1892
1358         Новосибирск              573             2006
1359               Минск             1115             2112
1360     Санкт-Петербург             2217             2851
1361              Москва             5864             5333

[1362 rows x 3 columns]

In [12]:
print(f'Наибольшее количество вакансий в стране: {df[df["vacancies_count"] == 0].iloc[0]["area_name"]}')

Наибольшее количество вакансий в стране: Россия


3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [ ]:
# текст запроса
query_5_3 = '''
SELECT 
    e.name AS employer,
    COUNT(DISTINCT v.area_id) AS regions_count
FROM employers e
LEFT JOIN vacancies v ON v.employer_id = e.id
GROUP BY e.name
ORDER BY regions_count DESC;
'''

In [14]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
df

employer  regions_count
0                      Яндекс            181
1                  Ростелеком            152
2                  Спецремонт            116
3      Поляков Денис Иванович             88
4                   ООО ЕФИН              71
...                       ...            ...
23170                  СДЕЛКА              0
23171            Alandr Group              0
23172                     СДК              0
23173      Lemon Land Lombard              0
23174                  Махина              0

[23175 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности.

In [ ]:
# текст запроса
query_5_4 = '''
SELECT 
    COUNT(*) AS employers_without_industry
FROM employers e
LEFT JOIN employers_industries ei ON e.id = ei.employer_id
WHERE ei.industry_id IS NULL;
'''

In [16]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
df

employers_without_industry
0                        8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности.

In [ ]:
# текст запроса
query_5_5 = '''
WITH employers_with_industries AS (
    SELECT 
        e.id,
        e.name,
        COUNT(ei.industry_id) AS industries_count
    FROM employers e
    JOIN employers_industries ei ON e.id = ei.employer_id
    GROUP BY e.id, e.name
    HAVING COUNT(ei.industry_id) = 4
)
SELECT name
FROM employers_with_industries
ORDER BY name
OFFSET 2
LIMIT 1;
'''

In [18]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
df

name
0  2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана «Разработка программного обеспечения».


In [ ]:
# текст запроса
query_5_6 = '''
SELECT 
    COUNT(DISTINCT e.id) AS software_dev_employers_count
FROM employers e
JOIN employers_industries ei ON e.id = ei.employer_id
JOIN industries i ON ei.industry_id = i.id
WHERE i.name = 'Разработка программного обеспечения';
'''

In [20]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
df

software_dev_employers_count
0                          3553

7. Для компании «Яндекс» выведите список [городов-миллионников](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8), в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку "Total" с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

    Если возникнут трудности с этим заданием, посмотрите материалы модуля 6.4 «Как получать данные из веб-источников и API».

In [13]:
import pandas as pd
import wikipedia

wikipedia.set_lang("ru")

# Получаем страницу
page = wikipedia.page("Города-миллионеры России")
html = page.html()
tables = pd.read_html(html)

# Берем первую таблицу
df = tables[0]

# Извлекаем только названия городов
cities_list = df['Город'].tolist()

# Выводим список городов
print("Список городов-миллионеров России:")

for i, city in enumerate(cities_list, 1):
    print(f"{i}. {city}")

# Преобразуем список городов в строку для SQL
cities_sql = ', '.join([f"'{city}'" for city in cities_list])

Список городов-миллионеров России:
1. Москва
2. Санкт-Петербург
3. Новосибирск
4. Екатеринбург
5. Казань
6. Красноярск
7. Нижний Новгород
8. Челябинск
9. Уфа
10. Краснодар
11. Самара
12. Ростов-на-Дону
13. Омск
14. Воронеж
15. Пермь
16. Волгоград


In [ ]:
# текст запроса
query_5_7 = f'''
WITH yandex_vacancies AS (
    SELECT 
        a.name AS city,
        COUNT(v.id) AS vacancies_count
    FROM vacancies v
    JOIN areas a ON v.area_id = a.id
    JOIN employers e ON v.employer_id = e.id
    WHERE e.name = 'Яндекс'
      AND a.name IN ({cities_sql})
    GROUP BY a.name
)
SELECT city, vacancies_count
FROM yandex_vacancies

UNION ALL

SELECT 'Total' AS city, SUM(vacancies_count) AS vacancies_count
FROM yandex_vacancies

ORDER BY vacancies_count ASC;
'''

In [21]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)
df

city  vacancies_count
0              Омск             21.0
1         Челябинск             22.0
2        Красноярск             23.0
3         Волгоград             24.0
4             Пермь             25.0
5            Казань             25.0
6    Ростов-на-Дону             25.0
7            Самара             26.0
8               Уфа             26.0
9         Краснодар             30.0
10          Воронеж             32.0
11      Новосибирск             35.0
12  Нижний Новгород             36.0
13     Екатеринбург             39.0
14  Санкт-Петербург             42.0
15           Москва             54.0
16            Total            485.0

***

### Выводы по анализу работодателей

- Лидером по количеству вакансий является Яндекс (1933), на пятом месте - Газпром нефть (331).

- Максимальное число работодателей без вакансий сосредоточено в России и Казахстане.

- Крупные компании, такие как Яндекс (181 регион) и Ростелеком (152 региона), публикуют вакансии сразу в многих регионах.

- 8419 работодателей не указали сферу деятельности, а 3553 компании относятся к разработке программного обеспечения.

- Вакансии «Яндекса» распределены по городам-миллионникам, лидируют Москва и Санкт-Петербург, всего 485 вакансий.


## 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

    Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова `'data'` или `'данн'`.

    *Обратите внимание, что названия вакансий могут быть написаны в любом регистре.*


In [ ]:
# текст запроса
query_6_1 = '''
SELECT 
    COUNT(*) AS data_related_vacancies
FROM vacancies
WHERE LOWER(name) LIKE '%data%' OR LOWER(name) LIKE '%данн%';
'''

In [26]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
df

data_related_vacancies
0                    1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
    * 'data scientist'
    * 'data science'
    * 'исследователь данных'
    * 'ML' (здесь не нужно брать вакансии по HTML)
    * 'machine learning'
    * 'машинн%обучен%'

    **В следующих заданиях мы продолжим работать с вакансиями по этому условию.**

    Считаем вакансиями для специалистов уровня Junior следующие:
    + в названии есть слово “junior” **или**
    + требуемый опыт — «Нет опыта» **или**
    + тип трудоустройства — «Стажировка».


In [ ]:
# текст запроса
query_6_2 = '''
SELECT 
    COUNT(*) AS junior_data_scientist_vacancies
FROM vacancies
WHERE (
        LOWER(name) LIKE '%data scientist%' OR
        LOWER(name) LIKE '%data science%' OR
        LOWER(name) LIKE '%исследователь данных%' OR
        LOWER(name) LIKE '%ml%' AND LOWER(name) NOT LIKE '%html%' OR
        LOWER(name) LIKE '%machine learning%' OR
        LOWER(name) LIKE '%машинн%обучен%'
      )
  AND (
        LOWER(name) LIKE '%junior%' OR
        experience = 'Нет опыта' OR
        employment = 'Стажировка'
      );
'''

In [30]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
df

junior_data_scientist_vacancies
0                               51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или Postgres?

    *Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [ ]:
# текст запроса
query_6_3 = f'''
SELECT 
    COUNT(*) AS ds_vacancies_with_sql
FROM vacancies
WHERE (
        LOWER(name) LIKE '%data scientist%' OR
        LOWER(name) LIKE '%data science%' OR
        LOWER(name) LIKE '%исследователь данных%' OR
        LOWER(name) LIKE '%ml%' AND LOWER(name) NOT LIKE '%html%' OR
        LOWER(name) LIKE '%machine learning%' OR
        LOWER(name) LIKE '%машинн%обучен%'
      )
AND (
        LOWER(key_skills) LIKE '%sql%'
     OR LOWER(key_skills) LIKE '%postgres%'
);
'''

In [36]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
df

ds_vacancies_with_sql
0                    229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS. Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

    *Это можно сделать помощью запроса, аналогичного предыдущему.*

In [40]:
# текст запроса
query_6_4 = f'''
SELECT 
    COUNT(*) AS ds_vacancies_with_sql
FROM vacancies
WHERE (
        LOWER(name) LIKE '%data scientist%' OR
        LOWER(name) LIKE '%data science%' OR
        LOWER(name) LIKE '%исследователь данных%' OR
        LOWER(name) LIKE '%ml%' AND LOWER(name) NOT LIKE '%html%' OR
        LOWER(name) LIKE '%machine learning%' OR
        LOWER(name) LIKE '%машинн%обучен%'
      )
AND LOWER(key_skills) LIKE '%python%';
'''

In [41]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
df

ds_vacancies_with_sql
0                    357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до **двух знаков после точки-разделителя**.

In [42]:
# текст запроса
query_6_5 = f'''
SELECT 
    ROUND(
        AVG(
            ARRAY_LENGTH(
                REGEXP_SPLIT_TO_ARRAY(key_skills, E'[\\t,;]'), 1
            )
        )::numeric,
        2
    ) AS avg_skills_count
FROM vacancies
WHERE
    (
        LOWER(name) LIKE '%data scientist%' OR
        LOWER(name) LIKE '%data science%' OR
        LOWER(name) LIKE '%исследователь данных%' OR
        LOWER(name) LIKE '%ml%' AND LOWER(name) NOT LIKE '%html%' OR
        LOWER(name) LIKE '%machine learning%' OR
        LOWER(name) LIKE '%машинн%обучен%'
    );
'''

In [43]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
df

avg_skills_count
0              6.63

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в среднем указывают для каждого типа требуемого опыта (уникальное значение из поля `experience`).

    При решении задачи примите во внимание следующее:
    1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
    2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
    3. Если в расчётах участвует `null`, в результате он тоже даст `null` (посмотрите, что возвращает запрос `select 1 + null`). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит `null` на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

    Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до **целого числа**.

In [48]:
# текст запроса
query_6_6 = f'''
SELECT
    experience,
    ROUND(
        AVG(
            (COALESCE(salary_from, 0) + COALESCE(salary_to, 0)) / 
            CASE 
                WHEN salary_from IS NOT NULL AND salary_to IS NOT NULL THEN 2
                ELSE 1
            END
        )
    )::integer AS avg_salary
FROM vacancies
WHERE 
    (
        LOWER(name) LIKE '%data scientist%' OR
        LOWER(name) LIKE '%data science%' OR
        LOWER(name) LIKE '%исследователь данных%' OR
        LOWER(name) LIKE '%ml%' AND LOWER(name) NOT LIKE '%html%' OR
        LOWER(name) LIKE '%machine learning%' OR
        LOWER(name) LIKE '%машинн%обучен%'
    )
    AND (salary_from IS NOT NULL OR salary_to IS NOT NULL)
GROUP BY experience
ORDER BY experience;
'''

In [49]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
df

experience  avg_salary
0         Более 6 лет      157933
1           Нет опыта       74643
2  От 1 года до 3 лет      144233
3       От 3 до 6 лет      256454

### Выводы по предметному анализу
- Всего 1771 вакансия имеют отношение к данным, что демонстрирует интерес работодателей к этой сфере.

- Для начинающих дата-сайентистов (Junior) найдена 51 вакансия, что показывает ограниченное количество предложений для новичков.

- Среди DS-вакансий навыки SQL или Postgres указаны в 229 позициях, а Python в 357 вакансиях, что подтверждает ключевую роль этих технологий в работе специалистов по данным.

- В среднем каждая DS-вакансия содержит 6,63 ключевых навыка, что говорит о высоких требованиях к квалификации и разнообразии используемых инструментов.

- Средняя предлагаемая зарплата зависит от опыта: дата-сайентист с опытом 3-6 лет может рассчитывать на 256454, с опытом более 6 лет - на 157933, с опытом 1-3 года - на 144233, а новичок без опыта - на 74643.

## Общий вывод по проекту

### Подведем итог исследования
Исследование показало, что база вакансий достаточно объёмная и разнообразная, что позволяет проводить статистически значимый анализ. Рынок труда сильно сконцентрирован в крупных городах, таких как Москва и Санкт-Петербург, при этом в некоторых регионах представлено всего несколько вакансий. Примерно у половины вакансий указана хотя бы одна граница зарплаты, а средняя предлагаемая зарплата существенно зависит от опыта: специалисты с 3–6 лет опыта могут рассчитывать на значительно более высокий доход, чем новички.

Основной формат работы - полный день с полной занятостью, но встречаются и удалённая работа, гибкий график, стажировки и проектная работа, что отражает разнообразие условий трудоустройства. Большинство вакансий рассчитано на специалистов с опытом от 1 до 6 лет, при этом есть предложения и для соискателей без опыта.

Лидерами по количеству вакансий являются крупные компании, такие как Яндекс и Газпром нефть. Значительная часть работодателей не указывает сферу деятельности. Крупные компании публикуют вакансии сразу в нескольких регионах.

В сегменте Data Science видно устойчивый интерес: 1771 вакансия связана с данными, но количество предложений для начинающих специалистов ограничено. Среди DS-вакансий наиболее востребованы навыки Python, SQL и Postgres, а каждая вакансия в среднем включает около семи ключевых навыков, что отражает высокие требования к квалификации специалистов.

В целом рынок вакансий демонстрирует концентрацию в мегаполисах, высокий уровень требований к специалистам и прямую зависимость уровня зарплаты от опыта и компетенций.

### Дополнительные исследования

In [ ]:
# текст запроса
query_7_1 = f'''
-- Топ-20 самых популярных ключевых навыков
SELECT skill, COUNT(*) AS occurrences
FROM (
    SELECT unnest(regexp_split_to_array(key_skills, E'[\\t,;]')) AS skill
    FROM vacancies
    WHERE key_skills IS NOT NULL
) AS skills
GROUP BY skill
ORDER BY occurrences DESC
LIMIT 20;
'''

In [59]:
# результат запроса
df = pd.read_sql_query(query_7_1, connection)
df

skill  occurrences
0                      SQL         6980
1                      Git         4733
2          Английский язык         4525
3         Работа в команде         3923
4                    Linux         3747
5               JavaScript         3628
6                   Python         3447
7     Управление проектами         2764
8      1С программирование         2647
9               PostgreSQL         2452
10          Грамотная речь         2298
11                    Java         2295
12  Аналитическое мышление         2295
13       1С: Предприятие 8         2234
14                     ООП         2075
15                    HTML         2022
16           MS PowerPoint         1964
17                   MySQL         1936
18                     PHP         1932
19         Пользователь ПК         1899

In [ ]:
# текст запроса
query_7_2 = f'''
-- Компании с наибольшим количеством уникальных вакансий
SELECT e.name AS employer, COUNT(*) AS vacancies_count
FROM vacancies v
JOIN employers e ON v.employer_id = e.id
GROUP BY e.name
ORDER BY vacancies_count DESC
LIMIT 20;
'''

In [61]:
# результат запроса
df = pd.read_sql_query(query_7_2, connection)
df

employer  vacancies_count
0                     Яндекс             1933
1                 Ростелеком              491
2                   Тинькофф              444
3                       СБЕР              428
4              Газпром нефть              331
5                ИК СИБИНТЕК              327
6                        МТС              292
7                    DataArt              247
8      Совкомбанк Технологии              204
9                 Первый Бит              176
10                      Ozon              175
11                       IBS              175
12                  Andersen              155
13                    Тензор              155
14                Спецремонт              151
15                        VK              150
16    Поляков Денис Иванович              127
17  Иннотех, Группа компаний              122
18           Bell Integrator              121
19                     ANCOR              118

In [ ]:
# текст запроса
query_7_3 = f'''
-- Средняя зарплата и количество вакансий по отраслям
SELECT 
    i.name AS industry,
    ROUND(AVG(
        (COALESCE(v.salary_from, 0) + COALESCE(v.salary_to, 0)) / 
        CASE 
            WHEN v.salary_from IS NOT NULL AND v.salary_to IS NOT NULL THEN 2 
            ELSE 1 
        END
    )) AS avg_salary,
    COUNT(v.id) AS vacancies_count
FROM vacancies v
JOIN employers e ON v.employer_id = e.id
LEFT JOIN employers_industries ei ON e.id = ei.employer_id
LEFT JOIN industries i ON ei.industry_id = i.id
WHERE v.salary_from IS NOT NULL OR v.salary_to IS NOT NULL
GROUP BY i.name
ORDER BY avg_salary DESC
LIMIT 20;
'''

In [65]:
# результат запроса
df = pd.read_sql_query(query_7_3, connection)
df

industry  avg_salary  \
0           Агрохимия (продвижение, оптовая торговля)    381429.0   
1                            Агрохимия (производство)    339875.0   
2                                      Игорный бизнес    191905.0   
3                   Корма для животных (производство)    168300.0   
4                                          Зоомагазин    167613.0   
5                                  Кадровые агентства    144530.0   
6       Лакокрасочная продукция, сырье (производство)    142803.0   
7   Бытовая химия, парфюмерия, косметика  (произво...    130120.0   
8   Лакокрасочная продукция, сырье (продвижение, о...    128445.0   
9   Дорожно-строительная техника, сельскохозяйстве...    122878.0   
10  Дорожно-строительная техника, сельскохозяйстве...    119919.0   
11  Оборудование для пищевой промышленности, упако...    119656.0   
12  Финансово-кредитное посредничество (биржа, бро...    117714.0   
13                Разработка программного обеспечения    117659.0   
14  Управляющая, инвестиционная компания (управлен...    114297.0   
15  Корма для животных (продвижение, оптовая торго...    113667.0   
16                          Ветеринарная деятельность    113364.0   
17                     Учреждение соц.помощи и защиты    113125.0   
18                                Лизинговые компании    112306.0   
19                                                НПФ    104000.0   

    vacancies_count  
0                14  
1                16  
2                20  
3                29  
4                31  
5               711  
6                26  
7                78  
8                32  
9                90  
10              106  
11              106  
12              192  
13             4572  
14               92  
15               57  
16               13  
17               16  
18               19  
19               13

In [ ]:
# текст запроса
query_7_4 = f'''
-- Средняя зарплата и количество вакансий по регионам
SELECT 
    a.name AS area,
    ROUND(AVG(
        (COALESCE(v.salary_from, 0) + COALESCE(v.salary_to, 0)) / 
        CASE 
            WHEN v.salary_from IS NOT NULL AND v.salary_to IS NOT NULL THEN 2 
            ELSE 1 
        END
    )) AS avg_salary,
    COUNT(v.id) AS vacancies_count
FROM vacancies v
JOIN areas a ON v.area_id = a.id
WHERE v.salary_from IS NOT NULL OR v.salary_to IS NOT NULL
GROUP BY a.name
ORDER BY avg_salary DESC
LIMIT 20;
'''

In [67]:
# результат запроса
df = pd.read_sql_query(query_7_4, connection)
df

area  avg_salary  vacancies_count
0                             Германия    452751.0                2
1                             Болгария    324348.0                3
2                              Испания    300000.0                1
3                           Черногория    288589.0                4
4                                Литва    263019.0                2
5                               Польша    257156.0                3
6   Петропавловка (Республика Бурятия)    250000.0                1
7                               Турция    235435.0               19
8                                 Кипр    223736.0               39
9                              Кутаиси    220475.0                3
10                Дзержинск (Беларусь)    220349.0                3
11                             Армения    219570.0              106
12                         Таджикистан    215094.0                2
13                      Великобритания    208962.0                2
14                              Сербия    208423.0               12
15                             Тбилиси    207156.0               80
16                             Эстония    204570.0                2
17                              Латвия    204570.0                2
18                              Шушары    200000.0                1
19    Павловск (Ленинградская область)    200000.0                1

### Выводы

Анализ ключевых навыков показал, что наиболее востребованы SQL, Git, Python, JavaScript, навыки коммуникации и управление проектами, что подчёркивает важность сочетания hard и soft навыков. Среди работодателей лидируют Яндекс, Ростелеком и Тинькофф.

Средняя зарплата по отраслям и регионам демонстрирует большой разброс: высокие зарплаты наблюдаются в агрохимической отрасли, игорном бизнесе и зарубежных регионах, а разработка программного обеспечения имеет высокое количество вакансий, но среднюю зарплату ниже топовых отраслей. Региональные данные показывают как концентрацию вакансий в крупных городах, так и высокооплачиваемые вакансии за рубежом. В целом рынок труда демонстрирует высокие требования к квалификации специалистов и прямую зависимость уровня зарплаты от опыта и навыков.

### Прогнозы

С учётом выявленных трендов можно ожидать, что спрос на специалистов с навыками Python, SQL и Git будет сохраняться или расти, особенно в IT и финансовом секторе. Крупные компании, такие как Яндекс и Ростелеком, вероятно, будут расширять набор персонала, в том числе в нескольких регионах одновременно.

Зарплатные прогнозы показывают, что специалисты с опытом работы от 3 до 6 лет будут получать значительное преимущество в оплате, а новичкам придётся конкурировать за ограниченное число вакансий. В перспективе рост зарплат возможен в высокодоходных отраслях (агрохимия, игорный бизнес) и зарубежных рынках, тогда как средние зарплаты в IT будут стабильными, но с большим числом вакансий.

### Планы дальнейшего исследования

- Провести анализ совпадения ключевых навыков с отраслью, чтобы выявить, какие навыки особенно востребованы в IT, финансах, промышленности и других секторах.

- Исследовать корреляцию между опытом, навыками и зарплатой, чтобы понять, какие комбинации дают максимальную оплату.

- Провести кластеризацию регионов и отраслей по средней зарплате и числу вакансий для выявления «дефицитных» или высокооплачиваемых сегментов.

- Проанализировать зарубежный рынок вакансий: средние зарплаты, востребованные навыки и отрасли, чтобы прогнозировать глобальные тренды.